### PR TITLE
remove workers global

### PR DESCRIPTION
**Pull request by rprimet**

### CODE REVIEW

File: dicodile/update_z/dicod.py [hunk 1]
```diff
@@ -125,29 +124,30 @@ def dicod(X_i, D, reg, z0=None, DtD=None, n_seg='auto', strategy='greedy',
         freeze_support=freeze_support, warm_start=warm_start
     )
 
-    comm = _spawn_workers(n_workers, hostfile)
-    t_transfert, workers_segments = _send_task(comm, X_i, D, z0, DtD, w_world,
-                                               params)
+    workers = _spawn_workers(n_workers, hostfile)
+    t_transfer, workers_segments = _send_task(workers.comm, X_i,
```
Comment (tomMoral) [inline dicodile/update_z/dicod.py:128]: Mayeb better to just pass the `workers` and take the `comm` in `send_command_to_workers`?

File: dicodile/update_z/distributed_sparse_encoder.py [hunk 2]
```diff
@@ -63,10 +61,10 @@ def init_workers(self, X, D_hat, reg, params, z0=None, DtD=None):
         self.params['precomputed_DtD'] = DtD is not None
         self.params['verbose'] = self.verbose
 
-        send_command_to_reusable_workers(constants.TAG_DICODILE_SET_TASK,
-                                         verbose=self.verbose)
+        self.workers.send_command(constants.TAG_DICODILE_SET_TASK,
+                                  verbose=self.verbose)
         self.t_init, self.workers_segments = _send_task(
-            self.comm, X, D_hat, z0, DtD, w_world, self.params
+            self.workers.comm, X, D_hat, z0, DtD, w_world, self.params
```
Comment (tomMoral) [inline dicodile/update_z/distributed_sparse_encoder.py:67]: ditto

File: dicodile/update_z/distributed_sparse_encoder.py [hunk 3]
```diff
@@ -78,70 +76,73 @@ def set_worker_D(self, D, DtD=None):
             raise ValueError("The pre-computed value DtD need to be passed "
                              "each time D is updated.")
 
-        send_command_to_reusable_workers(constants.TAG_DICODILE_SET_D,
-                                         verbose=self.verbose)
-        _send_D(self.comm, D, DtD)
+        self.workers.send_command(constants.TAG_DICODILE_SET_D,
+                                  verbose=self.verbose)
+        _send_D(self.workers.comm, D, DtD)
```
Comment (tomMoral) [inline dicodile/update_z/distributed_sparse_encoder.py:81]: ditto

File: dicodile/update_z/dicod.py [hunk 4]
```diff
@@ -199,9 +199,9 @@ def reconstruct_pobj(X, D, reg, _log, t_init, t_reduce, n_workers,
 
 
 def _spawn_workers(n_workers, hostfile):
-    comm = get_reusable_workers(n_workers, hostfile=hostfile)
-    send_command_to_reusable_workers(constants.TAG_WORKER_RUN_DICOD)
-    return comm
+    workers = get_reusable_workers(n_workers, hostfile=hostfile)
```
Comment (tomMoral) [inline dicodile/update_z/dicod.py:202]: Maybe better to use directly `Workers` here no?
```suggestion
    workers = Workers(n_workers, hostfile=hostfile)
```

File: dicodile/workers/reusable_workers.py [hunk 5]
```diff
@@ -33,47 +29,29 @@ def __init__(self, n_workers, hostfile):
 
     def __del__(self):
         if not self.shutdown:
-            shutdown_reusable_workers()
+            self.shutdown_reusable_workers()
```
Comment (tomMoral) [inline dicodile/workers/reusable_workers.py:32]: Maybe the name of the `Workers` class should be changed to `MPIWorkers`?

File: dicodile/workers/reusable_workers.py [hunk 6]
```diff
@@ -33,47 +29,29 @@ def __init__(self, n_workers, hostfile):
...
+
+    def send_command(self, tag, verbose=0):
```
Comment (tomMoral) [inline dicodile/workers/reusable_workers.py:34]: It would be nice to add some doc string to these methods?

**Review by tomMoral — COMMENTED** [review]
Overall, this looks good. I would maybe simply drop the `get_reusable_executor` as they are not reusable anymore and simply use `Workers`.

File: dicodile/update_z/dicod.py (hunk 1 above)
Comment (rprimet) [reply dicodile/update_z/dicod.py]: done :-)

Comment (tomMoral) [reply dicodile/update_z/dicod.py]: Much cleaner, I like it :)

File: dicodile/workers/reusable_workers.py (hunk 6 above)
Comment (rprimet) [reply dicodile/workers/reusable_workers.py:34]: yes, and also maybe drop the leading underscore from methods that are used outside of the class?

Comment (tomMoral) [reply dicodile/workers/reusable_workers.py:34]: Yes this seems like a nice idea.

Comment (rprimet) [reply dicodile/workers/reusable_workers.py:34]: Oh actually I was confused, there are a few of those in `dicod.py` but not in the `MPIWorkers` class.

**Review by tomMoral — COMMENTED** [review]
LGTM!

**Pull request merged**